### ISSUE TITLE
SMP branch sometimes hangs when the program terminates (deadlock during cleanup)

### ISSUE THREAD

**Issue by feeley** [gambit/gambit#221]

**Comment by gambiteer** [comment]
I'm not sure that this is the only problem.

I've been experimenting tonight with chud-parallel.scm and the latest smp branch and I get

```
heine:~/programs/gambit/gambit-smp> time gsi -:p1 chud-parallel.scm
(time (pi digits))
    46622 ms real time
    46620 ms cpu time (46300 user, 320 system)
    341 collections accounting for 330 ms real time (268 user, 32 system)
    8873424048 bytes allocated
    460464 minor faults
    no major faults
Chudnovsky's algorithm using binary splitting in Gambit Scheme: digits 10000000, real time: 46.6227765083313.
Last 5 digits 55897.
46.324u 0.356s 0:46.68 99.9%	0+0k 0+0io 0pf+0w
heine:~/programs/gambit/gambit-smp> time gsi -:p8 chud-parallel.scm
(time (pi digits))
    16213 ms real time
    82164 ms cpu time (80992 user, 1172 system)
    164 collections accounting for 144 ms real time (540 user, 72 system)
    1822585976 bytes allocated
    481362 minor faults
    no major faults
Chudnovsky's algorithm using binary splitting in Gambit Scheme: digits 10000000, real time: 16.21338987350464.
Last 5 digits 55897.
81.012u 1.212s 0:16.26 505.6%	0+0k 0+0io 0pf+0w
heine:~/programs/gambit/gambit-smp> time gsi -:p4 chud-parallel.scm
^C^C^Z
Suspended
heine:~/programs/gambit/gambit-smp> kill -9 %+
heine:~/programs/gambit/gambit-smp> 
[1]    Killed                        gsi -:p4 chud-parallel.scm
0.004u 0.000s 1:20.45 0.0%	0+0k 0+0io 0pf+0w
heine:~/programs/gambit/gambit-smp> time gsi -:p4 chud-parallel.scm
(time (pi digits))
    17334 ms real time
    52904 ms cpu time (52216 user, 688 system)
    199 collections accounting for 141 ms real time (368 user, 52 system)
    3809068592 bytes allocated
    581012 minor faults
    no major faults
Chudnovsky's algorithm using binary splitting in Gambit Scheme: digits 10000000, real time: 17.33468723297119.
Last 5 digits 55897.
52.248u 0.728s 0:17.40 304.3%	0+0k 0+0io 0pf+0w
```

Notice the time I had to kill it, it had only gotten .004 seconds into the computation before it hung (for over a minute before I killed it).

This was configured with

```
heine:~/programs/gambit/gambit-smp> gsi -v
v4.8.8 20170203122653 x86_64-unknown-linux-gnu "./configure 'CC=gcc -march=native -D___CAN_IMPORT_CLIB_DYNAMICALLY' '--enable-single-host' '--enable-multiple-versions' '--enable-shared' '--enable-multiple-threaded-vms' '--prefix=/usr/local/Gambit-smp'"
```

**Comment by feeley** [comment]
Fixed by commit 030aeb3f65c3c4c9eaefbbffa759e49079e40d4f
